src/core/pyproxy.c

```diff
@@ -13,6 +13,136 @@ _Py_IDENTIFIER(add_done_callback);
+ * protocol API defined in `abstract.c`. We wrote these tests to check whether
+ * the corresponding CPython APIs are likely to work without actually creating
+ * any temporary objects.
+ */
+int
+pyproxy_getflags(PyObject* pyobj)
+{
+  // Reduce casework by ensuring that protos aren't NULL.
+  PyTypeObject* obj_type = pyobj->ob_type;
+
+  PySequenceMethods null_seq_proto = { 0 };
+  PySequenceMethods* seq_proto =
+    obj_type->tp_as_sequence ? obj_type->tp_as_sequence : &null_seq_proto;
+
+  PyMappingMethods null_map_proto = { 0 };
+  PyMappingMethods* map_proto =
+    obj_type->tp_as_mapping ? obj_type->tp_as_mapping : &null_map_proto;
+
+  PyAsyncMethods null_async_proto = { 0 };
+  PyAsyncMethods* async_proto =
+    obj_type->tp_as_async ? obj_type->tp_as_async : &null_async_proto;
+
+  PyBufferProcs null_buffer_proto = { 0 };
+  PyBufferProcs* buffer_proto =
+    obj_type->tp_as_buffer ? obj_type->tp_as_buffer : &null_buffer_proto;
+
+  int result = 0;
+  // PyObject_Size
+  if (seq_proto->sq_length || map_proto->mp_length) {
+    result |= HAS_LENGTH;
+  }
+  // PyObject_GetItem
+  if (map_proto->mp_subscript || seq_proto->sq_item) {
+    result |= HAS_GET;
+  } else if (PyType_Check(pyobj)) {
+    _Py_IDENTIFIER(__class_getitem__);
+    if (_PyObject_HasAttrId(pyobj, &PyId___class_getitem__)) {
+      result |= HAS_GET;
+    }
+  }
+  // PyObject_SetItem
+  if (map_proto->mp_ass_subscript || seq_proto->sq_ass_item) {
+    result |= HAS_SET;
+  }
+  // PySequence_Contains
+  if (seq_proto->sq_contains) {
+    result |= HAS_CONTAINS;
+  }
+  // PyObject_GetIter
+  if (obj_type->tp_iter || PySequence_Check(pyobj)) {
+    result |= IS_ITERABLE;
+  }
+  if (PyIter_Check(pyobj)) {
+    result &= ~IS_ITERABLE;
+    result |= IS_ITERATOR;
+  }
+  // There's no CPython API that corresponds directly to the "await" keyword.
+  // Looking at disassembly, "await" translates into opcodes GET_AWAITABLE and
+  // YIELD_FROM. GET_AWAITABLE uses _PyCoro_GetAwaitableIter defined in
+  // genobject.c. This tests whether _PyCoro_GetAwaitableIter is likely to
+  // succeed.
+  if (async_proto->am_await || gen_is_coroutine(pyobj)) {
+    result |= IS_AWAITABLE;
+  }
+  if (buffer_proto->bf_getbuffer) {
+    result |= IS_BUFFER;
+  }
+  // PyObject_Call (from call.c)
+  if (_PyVectorcall_Function(pyobj) || PyCFunction_Check(pyobj) ||
+      obj_type->tp_call) {
+    result |= IS_CALLABLE;
+  }
+  return result;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Object protocol wrappers
+//
+// This section defines wrappers for Python Object protocol API calls that we
+// are planning to offer on the PyProxy. Much of this could be written in
+// Javascript instead. Some reasons to do it in C:
+//  1. Some CPython APIs are actually secretly macros which cannot be used from
+//     Javascript.
+//  2. The code is a bit more concise in C.
+//  3. It may be preferable to minimize the number of times we cross between
```

Comment:
IMHO we should avoid talking about "not being sure" about something in the code. I think the solution as it is now has its right to exist, therefore it is fine.

Reply:
@hoodmane what do you think about this?

Reply:
I'm happy to change the comment. Should I just remove "I'm not sure"?

Reply:
In general I think it is fine to have comments stating things like "I'm not sure". For reference, CPython 3.8.2 source has 27 comments saying variations of "I'm not sure", the emscripten source has 11.